### PR TITLE
feat: create kubernetes events in case of insufficient capacity errors

### PR DESCRIPTION
**Pull request by Fedosin**
**Description**

We need to have improved visibility of ICE events when karpenter encounters them.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/machine/lifecycle/launch.go [hunk 1]
```diff
@@ -104,6 +106,13 @@ func (l *Launch) launchMachine(ctx context.Context, machine *v1alpha5.Machine) (
 	created, err := l.cloudProvider.Create(ctx, machine)
 	if err != nil {
 		if !cloudprovider.IsInsufficientCapacityError(err) {
+			l.recorder.Publish(events.Event{
```
Comment (bwagner5) [inline pkg/controllers/machine/lifecycle/launch.go:109]: this block is executed when the error was NOT an InsufficientCapacityError:

```
		if !cloudprovider.IsInsufficientCapacityError(err) {
```

File: pkg/controllers/machine/lifecycle/launch.go [hunk 2]
```diff
@@ -107,6 +109,13 @@ func (l *Launch) launchMachine(ctx context.Context, machine *v1alpha5.Machine) (
 			machine.StatusConditions().MarkFalse(v1alpha5.MachineLaunched, "LaunchFailed", truncateMessage(err.Error()))
 			return nil, fmt.Errorf("creating machine, %w", err)
 		}
+		l.recorder.Publish(events.Event{
```
Comment (jackfrancis) [inline pkg/controllers/machine/lifecycle/launch.go:112]: Should we convert this flow into a switch case on the error type and then default to the error return if we don't recognize it? That structure would make adding more error-specific events in the future easier.

Comment (bwagner5) [reply pkg/controllers/machine/lifecycle/launch.go:112]: Yeah, I think that flow would be a lot easier to reason about. It took me a second to catch that the if was `Not an InsufficientCapacityError`

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀

File: pkg/controllers/machine/lifecycle/launch.go (hunk 2 above)
Comment (Fedosin) [reply pkg/controllers/machine/lifecycle/launch.go]: Fixed :) I added a `switch` to check the error type.

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀

**Review by njtran — APPROVED** [review]
LGTM!

**Pull request merged**